### PR TITLE
GROOVY-10056: Inferred parameter type of lambda expression for multi-…

### PR DESCRIPTION
**Pull request by danielsun1106**
…dimensions array is not correct

https://issues.apache.org/jira/browse/GROOVY-10056

This may be the minimum change for the issue.

### CODE REVIEW

File: src/main/java/org/codehaus/groovy/transform/stc/StaticTypeCheckingVisitor.java [hunk 1]
```diff
@@ -5305,10 +5305,7 @@ protected ClassNode inferReturnTypeGenerics(final ClassNode receiver, final Meth
                                             applyGenericsContextToParameterClass(resolvedPlaceholders, paramType));
                         }
                     }
-                    if (isVargs && lastArg && argumentType.isArray()) {
-                        argumentType = argumentType.getComponentType();
-                    }
-                    if (isVargs && lastArg && paramType.isArray()) {
+                    if (isVargs && lastArg && paramType.isArray() && !argumentType.isArray()) {
                         paramType = paramType.getComponentType();
                     }
```
Comment (eric-milles) [inline src/main/java/org/codehaus/groovy/transform/stc/StaticTypeCheckingVisitor.java:5310]: What about tying the checks together?  I have tried this for the test case, but have run out of time today to run the full test suite.
```java
                    if (isVargs && lastArg && argumentType.isArray() && paramType.isArray()) {
                        argumentType = argumentType.getComponentType();
                        paramType = paramType.getComponentType();
                    }
```

Comment (danielsun1106) [reply src/main/java/org/codehaus/groovy/transform/stc/StaticTypeCheckingVisitor.java:5310]: As the following code will make `argumentType` and `paramType` non-array type, so `isVargs && lastArg && argumentType.isArray() && paramType.isArray()` will be always `false` IMO.

https://github.com/apache/groovy/blob/master/src/main/java/org/codehaus/groovy/transform/stc/StaticTypeCheckingVisitor.java#L5297-L5300

Anyway, let me try according to you suggestion.

Comment (eric-milles) [reply src/main/java/org/codehaus/groovy/transform/stc/StaticTypeCheckingVisitor.java:5310]: So probably not correct to reduce the arrays together.  But I don't think it is correct to reduce the parameter type only in the argument is not still an array.  There are more complete variadic checks in other blocks; I think the ones in `inferReturnTypeGenerics` are quite simplistic.  Like if the param type is `Object[]`, it accepts any array type.

I did not get the error for "new ListExpression(Arrays.asList(..." and I'm not sure why that is.

Comment (danielsun1106) [reply src/main/java/org/codehaus/groovy/transform/stc/StaticTypeCheckingVisitor.java:5310]: > There are more complete variadic checks in other blocks; I think the ones in inferReturnTypeGenerics are quite simplistic.

Yep. The variadic checks here is not complete.

> So probably not correct to reduce the arrays together.

The cause of GROOVY-10056 and the above error `new ListExpression(Arrays.asList(...` is reducing type array not correctly.
```java
                    if (isVargs && lastArg && argumentType.isArray()) {
                        argumentType = argumentType.getComponentType();
                    }
                    if (isVargs && lastArg && paramType.isArray()) {
                        paramType = paramType.getComponentType();
                    }
```
(https://github.com/apache/groovy/blob/master/src/main/java/org/codehaus/groovy/transform/stc/StaticTypeCheckingVisitor.java#L5310-L5315)

> I did not get the error for "new ListExpression(Arrays.asList(..." and I'm not sure why that is.

You could try run tests with the gradle option `--no-build-cache`.

To be honest, I have no better idea to fix GROOVY-10056 for now.

**Pull request merged**